sound/soc/codecs/rt711-sdca.c

```diff
@@ -0,0 +1,1309 @@
+		&owner);
+	if (ret < 0)
+		return 0;
+
+	/* if owner is device then there is no button event from device */
+	if (owner == 1)
+		return 0;
+
+	/* read UMP message offset */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_HID, RT711_SDCA_ENT_HID01, RT711_SDCA_CTL_HIDTX_MESSAGE_OFFSET, 0),
+		&offset);
+	if (ret < 0)
+		goto _end_btn_det_;
+
+	for (idx = 0; idx < sizeof(buf); idx++) {
+		ret = regmap_read(rt711->regmap,
+			RT711_BUF_ADDR_HID1 + offset + idx, &val);
+		if (ret < 0)
+			goto _end_btn_det_;
+		buf[idx] = val & 0xff;
+	}
+
+	if (buf[0] == 0x11) {
+		switch (buf[1] & 0xf0) {
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+		switch (buf[2]) {
+		case 0x01:
+		case 0x10:
+			btn_type |= SND_JACK_BTN_0;
+			break;
+		case 0x02:
+		case 0x20:
+			btn_type |= SND_JACK_BTN_1;
+			break;
+		case 0x04:
+		case 0x40:
+			btn_type |= SND_JACK_BTN_2;
+			break;
+		case 0x08:
+		case 0x80:
+			btn_type |= SND_JACK_BTN_3;
+			break;
+		}
+	}
+
+_end_btn_det_:
+	/* Host is owner, so set back to device */
+	if (owner == 0)
+		/* set owener to device */
+		regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUN_HID, RT711_SDCA_ENT_HID01,
+			RT711_SDCA_CTL_HIDTX_SET_OWNER_TO_DEVICE, 0), 0x01);
+
+	return btn_type;
+}
+
+static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
+{
+	unsigned int det_mode;
+	int ret;
+
+	/* get detected_mode */
+	ret = regmap_read(rt711->regmap,
+		SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_GE49, RT711_SDCA_CTL_DETECTED_MODE, 0),
+		&det_mode);
+	if (ret < 0)
+		goto io_error;
+
+	switch (det_mode) {
+	case 0x00:
+		rt711->jack_type = 0;
+		break;
+	case 0x03:
+		rt711->jack_type = SND_JACK_HEADPHONE;
+		break;
+	case 0x05:
+		rt711->jack_type = SND_JACK_HEADSET;
+		break;
+	}
+
+	/* write selected_mode */
+	if (det_mode) {
+		ret = regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_GE49, RT711_SDCA_CTL_SELECTED_MODE, 0),
+			det_mode);
+		if (ret < 0)
+			goto io_error;
+	}
+
+	dev_dbg(&rt711->slave->dev,
+		"%s, detected_mode=0x%x\n", __func__, det_mode);
+
+	return 0;
+
+io_error:
+	pr_err_ratelimited("IO error in %s, ret %d\n", __func__, ret);
+	return ret;
+}
+
+static void rt711_sdca_jack_detect_handler(struct work_struct *work)
+{
+	struct rt711_sdca_priv *rt711 =
+		container_of(work, struct rt711_sdca_priv, jack_detect_work.work);
+	int btn_type = 0, ret;
+	unsigned int sdca_stat0, sdca_stat8;
+
+	if (!rt711->hs_jack)
+		return;
+
+	if (!rt711->component->card->instantiated)
+		return;
+
+	/* SDW_SCP_SDCA_INT_SDCA_0 is used for jack detection */
+	ret = regmap_read(rt711->regmap, SDW_SCP_SDCA_INT1, &sdca_stat0);
+	if (ret < 0)
+		goto io_error;
+	if (sdca_stat0) {
+		ret = regmap_write(rt711->regmap, SDW_SCP_SDCA_INT1, sdca_stat0);
```

Comment:
that clears the SDCA interrupt, yes?

I am wondering if this is the right thing to do. The expectation was that the interrupt was cleared in the interrupt callback, but here this happens in a work queue, so it could be that the interrupt callback is called multiple times because the interrupt is still not cleared. 

Actually now that I write this I realize the SDCA code is not good enough, we need to re-read the SDCA bit in the bus loop before checking if there are SDCA interrupts left https://github.com/thesofproject/linux/blob/e1a25077acf02d83e034d27c814ef9a68a71b46f/drivers/soundwire/bus.c#L1571

What we have now is not good enough, there's a risk of losing interrupts (known SoundWire 'feature')